R/Spectra-functions.R

```diff
@@ -153,3 +153,104 @@ applyProcessing <- function(object, f = dataStorage(object),
         FALSE
     } else TRUE
 }
+
+#' @title Spectrum comparison
+#'
+#' @description
+#'
+#' Compare each spectrum in `x` with each spectrum in `y` with function `FUN`.
+#'
+#' @note
+#'
+#' Results might be slightly different if `x` and `y` are switched, because of
+#' the matching of the peaks by their m/z between spectra. Matching is performed
+#' by applying a `tolerance` and `ppm` to the second spectrum (i.e. from `y`)
+#' and, especially for `ppm`, the maximal accepted difference depend thus on
+#' the m/z values from the second spectrum.
+#'
+#' @param x [Spectra()]
+#'
+#' @param y [Spectra()]
+#'
+#' @param FUN `function` to compare the (m/z matched) intensities from each
+#'     spectrum in `x` with those in `y`.
+#'
+#' @param tolerance `numeric(1)` allowing to define a constant maximal accepted
+#'     difference between m/z values for peaks to be matched.
+#'
+#' @param ppm `numeric(1)` allowing to define a relative, m/z-dependent,
+#'     maximal accepted difference between m/z values for peaks to be matched.
+#'
+#' @param ... additional parameters passed to `FUN`.
+#'
+#' @return
+#'
+#' `matrix` with number of rows equal to length of `x` and number of columns
+#' equal `length(y)`.
+#'
+#' @examples
+#'
+#' library(Spectra)
+#' fl <- system.file("TripleTOF-SWATH/PestMix1_SWATH.mzML", package = "msdata")
+#' sps <- Spectra(fl, source = MsBackendMzR())
+#'
+#' sps <- sps[1:10]
+#' res <- .compare_spectra(sps, sps, ppm = 20)
+#' res <- .compare_spectra(sps, sps, method = "spearman", ppm = 40)
+#'
+#' res <- .compare_spectra(sps[1], sps)
+#' res <- .compare_spectra(sps, sps[1])
+#' @noRd
+.compare_spectra <- function(x, y = NULL, FUN = cor, tolerance = 0,
+                             ppm = 20, ...) {
+    x_idx <- seq_along(x)
+    y_idx <- seq_along(y)
+    mat <- matrix(NA_real_, nrow = length(x_idx), ncol = length(y_idx),
+                  dimnames = list(spectraNames(x), spectraNames(y)))
+    ## Might need some tuning - bplapply?
+    for (i in x_idx) {
```

Comment:
Do we really want to compare each spectrum against each other twice (I know that `.peaks_compare_intensities(x, y)` could be different to `.peaks_compare_intensities(y, x)` depending on the output of `closest`). Here we could avoid half of the computation by just copying the upper triangle of the matrix to the lower (and don't calculate the diagonal at all).
I guess `combn` could help here.

Reply:
I don't think we are calculating things twice. `x` and `y` are independent vectors (presumably not containing the same spectra). We are comparing each element in `x` with each element in `y`, but not *vice-versa*.

There is a second function for *self*comparison (when `y` is missing) that uses `combn` and copying the lower triangle.

Reply:
Ok, I missed that. I thought `x` and `y` can contain the same spectra (but they can if the user doesn't take care of it). The comparison of `x[i]` vs `y[j]` is independent so we can do it in parallel. It is not part of the processing queue so we have to do it "manually" here, or?

Reply:
Yep, we have to do it manually. I will try but my guess is that parallel processing will not be faster because of the required splitting/merging of the input.